File: examples/java/patterns/saga-pattern/saga-pattern.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import software.amazon.lambda.durable.DurableContext;
import software.amazon.lambda.durable.DurableHandler;
import software.amazon.lambda.durable.StepContext;

public class Saga extends DurableHandler<Map<String, Object>, Map<String, Object>> {

    // Compensation holder
    static class Compensation {
        final String name;
        final Runnable fn;

        Compensation(String name, Runnable fn) {
            this.name = name;
            this.fn = fn;
        }
    }

    @Override
    public Map<String, Object> handleRequest(Map<String, Object> event, DurableContext context) {
        List<Compensation> compensations = new ArrayList<>();

        try {
            String orderId = (String) event.get("order_id");
            double amount  = ((Number) event.get("amount")).doubleValue();

            // Step 1: Reserve inventory
            Map<String, Object> reservation = context.step(
                "reserve-inventory", Map.class,
                (StepContext ctx) -> {
                    ctx.getLogger().info("Reserving inventory for order: " + orderId);
                    return reserveInventory(orderId);
                }
            );
            compensations.add(new Compensation(
                "cancel-reservation",
                () -> cancelReservation((String) reservation.get("id"))
            ));

            // Step 2: Charge payment
            Map<String, Object> payment = context.step(
                "charge-payment", Map.class,
                (StepContext ctx) -> {
                    ctx.getLogger().info("Charging payment: " + amount);
                    return chargePayment(amount);
                }
            );
            compensations.add(new Compensation(
                "refund-payment",
                () -> refundPayment((String) payment.get("id"))
            ));

            // Step 3: Create shipment (no compensation because it's the last step)
            context.step(
                "create-shipment", Map.class,
                (StepContext ctx) -> {
                    ctx.getLogger().info("Creating shipment for order: " + orderId);
                    return createShipment(orderId);
                }
            );

            context.getLogger().info("Order completed successfully: " + orderId);
            return Map.of("success", true);

        } catch (Exception error) {
            context.getLogger().error("Order failed, running compensations: " + error.getMessage());

            // Run compensations in reverse to undo completed steps
            for (int i = compensations.size() - 1; i >= 0; i--) {
                Compensation comp = compensations.get(i);
                try {
                    context.step(comp.name, Void.class, (StepContext ctx) -> {
                        comp.fn.run();
                        return null;
                    });
                } catch (Exception compError) {
                    // Log but continue to attempt all compensations even if one fails
                    context.getLogger().error("Compensation failed: " + comp.name + " - " + compError.getMessage());
                }
            }

            throw error;
        }
    }

    // Mock APIs for demonstration purposes

    private Map<String, String> reserveInventory(String orderId) {
        return Map.of("id", "RES-" + orderId);
    }

    private void cancelReservation(String reservationId) {
        System.out.println("Reservation " + reservationId + " cancelled");
    }

    private Map<String, String> chargePayment(double amount) {
        return Map.of("id", "PAY-" + amount);
    }

    private void refundPayment(String paymentId) {
        System.out.println("Payment " + paymentId + " refunded");
    }

    private Map<String, String> createShipment(String orderId) {
        return Map.of("tracking_id", "TRACK-" + orderId);
    }
}

File: examples/typescript/patterns/saga-pattern/saga-pattern.ts
import { withDurableExecution, DurableContext } from '@aws/durable-execution-sdk-js';

export const handler = withDurableExecution(async (event: any, context: DurableContext) => {
  
  const compensations: Array<{ name: string; fn: () => Promise<void> }> = [];

  try {
    // Forward steps: each registers a compensation on success
    const reservation = await context.step('reserve-inventory', async () =>
      reserveInventory(event.orderId)
    );
    compensations.push({
      name: 'cancel-reservation',
      fn: () => cancelReservation(reservation.id)
    });

    const payment = await context.step('charge-payment', async () =>
      chargePayment(event.amount)
    );
    compensations.push({
      name: 'refund-payment',
      fn: () => refundPayment(payment.id)
    });

    await context.step('create-shipment', async () =>
      createShipment(event.orderId)
    );

    return { success: true };

  } catch (error) {
    // Run compensations in reverse to undo completed steps
    for (const comp of compensations.reverse()) {
      try {
        await context.step(comp.name, async () => comp.fn());
      } catch (compError) {
        context.logger.error(`Compensation failed: ${comp.name}`, compError);
      }
    }
    throw error;
  }
});

// Mock APIs for demonstration purposes

async function reserveInventory(orderId: string): Promise<{ id: string }> {
  return { id: `RES-${orderId}` };
}

async function cancelReservation(reservationId: string): Promise<void> {
  console.log(`Reservation ${reservationId} cancelled`);
}

async function chargePayment(amount: number): Promise<{ id: string }> {
  return { id: `PAY-${amount}` };
}

async function refundPayment(paymentId: string): Promise<void> {
  console.log(`Payment ${paymentId} refunded`);
}

async function createShipment(orderId: string): Promise<{ trackingId: string }> {
  return { trackingId: `TRACK-${orderId}` };
}
